src/test/testDirectives.ts

```diff
@@ -0,0 +1,505 @@
+          assert.strictEqual(details.objectType, mutationObjectType);
+          assert.strictEqual(details.field.args[0], arg);
+        }
+      },
+
+      enumTypeDirective: class extends SchemaDirectiveVisitor {
+        public visitEnum(enumType: GraphQLEnumType) {
+          assert.strictEqual(this.visitedType, enumType);
+          assert.strictEqual(enumType.name, 'Gender');
+          enumObjectType = enumType;
+        }
+      },
+
+      enumValueDirective: class extends SchemaDirectiveVisitor {
+        public visitEnumValue(value: GraphQLEnumValue, details: {
+          enumType: GraphQLEnumType,
+        }) {
+          assert.strictEqual(this.visitedType, value);
+          assert.strictEqual(value.name, 'NONBINARY');
+          assert.strictEqual(value.value, 'NONBINARY');
+          assert.strictEqual(details.enumType, enumObjectType);
+        }
+      },
+
+      inputTypeDirective: class extends SchemaDirectiveVisitor {
+        public visitInputObject(object: GraphQLInputObjectType) {
+          inputObjectType = object;
+          assert.strictEqual(this.visitedType, object);
+          assert.strictEqual(object.name, 'PersonInput');
+        }
+      },
+
+      inputFieldDirective: class extends SchemaDirectiveVisitor {
+        public visitInputFieldDefinition(field: GraphQLInputField, details: {
+          objectType: GraphQLInputObjectType,
+        }) {
+          assert.strictEqual(this.visitedType, field);
+          assert.strictEqual(field.name, 'name');
+          assert.strictEqual(details.objectType, inputObjectType);
+        }
+      }
+    });
+  });
+
+  it('can check if a visitor method is implemented', () => {
+    class Visitor extends SchemaVisitor {
+      public notVisitorMethod() {
+        return false;
+      }
+
+      public visitObject() {
+        return true;
+      }
+    }
+
+    const visitor = new Visitor;
+
+    assert.strictEqual(
+      Visitor.implementsVisitorMethod('notVisitorMethod'),
+      visitor.notVisitorMethod(),
+    );
+
+    assert.strictEqual(
+      Visitor.implementsVisitorMethod('visitObject'),
+      visitor.visitObject(),
+    );
+
+    assert.strictEqual(
+      Visitor.implementsVisitorMethod('visitInputFieldDefinition'),
+      false,
+    );
+
+    assert.strictEqual(
+      Visitor.implementsVisitorMethod('visitBogusType'),
+      false,
+    );
+  });
+
+  it('can use visitSchema for simple visitor patterns', () => {
+    class SimpleVisitor extends SchemaVisitor {
+      public visitCount = 0;
+      public names: string[] = [];
+
+      constructor(s: GraphQLSchema) {
+        super();
+        this.schema = s;
+      }
+
+      public visit() {
+        // More complicated visitor implementations might use the
+        // visitorSelector function more selectively, but this SimpleVisitor
+        // class always volunteers itself to visit any schema type.
+        visitSchema(this.schema, () => [this]);
```

Comment:
This test demonstrates how easy it is to use the `visitSchema` function even if you don't need to do something as complicated as what `SchemaDirectiveVisitor` does.